Revision of the API for visiting and changing AST

adams85 opened the pull request:
This PR is a follow-up to https://github.com/sebastienros/esprima-dotnet/pull/243. It aims to improve the API provided by the library for visiting and changing the AST.

* Reverts `AstVisitor` to its original design to keep the visitor API as straightforward as possible.
* In view of the upcoming major version, eliminates a few old anomalies of the `AstVisitor` API which have been kept delaying as requiring breaking changes.
* Changes the return type of `Visit` to `object?` to enable maximum freedom for inheritors. Thus the AST can even be transformed to other data structures than the Esprima AST.
* Due to this change, the term 'converting the AST' may include more general ideas than what `AstConverter` does. The term 'rewriting the AST' describes that better as `AstConverter` is designed for changing (rewriting) the AST. If a consumer wanted to transform (convert) the AST into another data structure, they would need to subclass `AstVisitor` because `AstConverter` would be useless in this case as no code could be (re)used from its implementation. So for this reason I suggest renaming `AstConverter` to `AstRewriter`. (However, I don't insist on this. If you disagree, we can undo it.)
* Changes `UpdateXXX` methods of `AstConverter` into extension methods (see `NodeExtensions`). (We can place them in the AST node classes as it's done in the case of .NET expression trees if you don't like extension methods.)

@Xicy @CharlieEriksen Could you adapt your existing pieces of code and test them with this proposal? Your feedback would be very valuable. Just let me know if you don't like or would change something.

There are some minor issues left to sort out, so please don't merge yet. More to come soon.

Xicy commented:
Thank you. It's working like a charm 🎉 

>@Xicy @CharlieEriksen Could you adapt your existing pieces of code and test them with this proposal? Your feedback would be very valuable. Just let me know if you don't like or would change something.

I checked out the branch and tested with Jsx to React converter it's working and I did not see any missed implementation. If I missed something could you please tell me. 

>We can place them in the AST node classes as it's done in the case of .NET expression trees if you don't like extension methods.

Maybe its([ConditionalExpression.cs Update method](https://github.com/dotnet/runtime/blob/v6.0.5/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/ConditionalExpression.cs#L81-L98)) be more elegant and easy to understand for future.

adams85 commented:
@Xicy Thanks for the feedback! I'm glad it works well for you but the major part of the credit goes to you! I've just refactored your code and polished off some rough edges.

You're right, AST node classes are a better place for `Update` methods, so I moved them there. However, I had to rename them to `UpdateWith` because there's a property called name Update in `ForStatement` already. Bummer.

@lahma I think the PR is now ready for review. I plan to submit two another PRs after this gets merged: I found some minor inconsistencies in the AST and I'd like better logical separation between the standard JS parser and JSX.

lahma reviewed (approved):
> However, I had to rename them to UpdateWith because there's a property named Update in ForStatement already. Bummer...

Well, you might consider to rename for statement's property, as long as estree output is the same `"update: "`. MDN seems to call the construct `for ([initialization]; [condition]; [final-expression])`.

> I plan to submit two more PRs after this gets merged: I found some minor inconsistencies in the AST and I'd like better logical separation between the standard JS parser and JSX.

Great! Like I mentioned before, JSX stuff inherits from JavaScriptParser on JS esprima side.

lahma commented:
Great collaboration folks, keep 'em coming! 🚀

adams85 commented on src/Esprima/Ast/ArrowFunctionExpression.cs:
```diff
@@ -34,9 +34,19 @@ public sealed class ArrowFunctionExpression : Expression, IFunction
+
+        public ArrowFunctionExpression UpdateWith(in NodeList<Expression> parameters, Node body)
+        {
+            if (NodeList.AreSame(parameters, Params) && body == Body)
```
I considered `ReferenceEquals` but I think in this case we don't have to require reference equality. If comparands happen to be differenct instances but describes the same node (value equality), then the original node already contains the same information, so we'd gain nothing with a new instance.

adams85 commented:
> Well, you might consider to rename for statement's property, as long as estree output is the same "update: ". MDN seems to call the construct for ([initialization]; [condition]; [final-expression]).

I think it's best to leave it as is now. We should avoid breaking changes which are not essential. And maybe the name `UpdateWith` is also a bit more expressive anyway. :)
